### PR TITLE
planner, variable: add test and ga the SET_VAR hint

### LINKED ISSUE #45892: Add tests to GA the SET_VAR hint
## Enhancement
Make sure it can take effect as what we want

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/45892

Problem Summary:

### What is changed and how it works?

Adding tests to test that the hint can take effect. And the vars can be restored after the SQL is executed.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: sessionctx/variable/setvar_affect.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	"sql_mode":                {},
+	"max_execution_time":      {},
+}
+
+func setHintUpdatable(vars []*SysVar) {
+	for _, v := range vars {
+		if _, ok := isHintUpdatable[v.Name]; ok {
+			v.IsHintUpdatable = true
+			if v.Type == TypeStr || v.Type == TypeEnum || v.Type == TypeTime || v.Type == TypeDuration {
+				fmt.Println(v.Name)
```
Comment (AilinKid) [inline sessionctx/variable/setvar_affect.go:112]: debug info？

File: sessionctx/variable/variable.go [hunk 2]
```diff
@@ -616,6 +616,11 @@ func GetSysVars() map[string]*SysVar {
 
 func init() {
 	sysVars = make(map[string]*SysVar)
+	setHintUpdatable(defaultSysVars)
+	// Clean the map
+	for k := range isHintUpdatable {
+		delete(isHintUpdatable, k)
```
Comment (hawkingrei) [inline sessionctx/variable/variable.go:622]: ```maps.Cleaer```

File: sessionctx/variable/setvar_affect.go (hunk 1 above)
Comment (winoros) [reply sessionctx/variable/setvar_affect.go:112]: Yes, it's used for generating test cases.

File: sessionctx/variable/setvar_test.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		varName string
+		cases   []string
+	}{
+		{"tidb_read_consistency", []string{"strict", "weak"}},
+		{"tidb_isolation_read_engines", []string{"tidb,tiflash", "tikv,tidb"}},
+		{"tidb_replica_read", []string{"follower", "prefer-leader"}},
+		{"tidb_partition_prune_mode", []string{"static", "dynamic"}},
+		{"mpp_version", []string{"0", "1"}},
+		{"tidb_opt_fix_control", []string{"44262:ON", "44389:ON,44823:ON"}},
+		// {"tidb_runtime_filter_type", []string{}},
```
Comment (elsa0520) [inline sessionctx/variable/setvar_test.go:126]: remove it ~

File: sessionctx/variable/setvar_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		{"tidb_runtime_filter_mode", []string{"OFF", "LOCAL"}},
+		{"sql_mode", []string{"", "ONLY_FULL_GROUP_BY"}},
+	}
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	for _, c := range testCases {
+		for _, setv := range c.cases {
+			tk.MustQuery(fmt.Sprintf("select /*+ set_var(%v=\"%v\") */ @@%v", c.varName, setv, c.varName)).Check(testkit.Rows(setv))
```
Comment (elsa0520) [inline sessionctx/variable/setvar_test.go:134]: It seems that this UT only test the ```set_var``` does not affect the variable. But it does not test whether the variable takes effect in the query by ```set_var```.
So wouldn't it be better to add such checks?

Comment (winoros) [reply sessionctx/variable/setvar_test.go:134]: There's no better idea to test each of them at one time. I think we can only pick some of them to test in the way you said.

File: sessionctx/variable/setvar_test.go (hunk 3 above)
Comment (winoros) [reply sessionctx/variable/setvar_test.go:126]: Or I make it not updatable this time?
Leave it for you once you support other types

Comment (elsa0520) [reply sessionctx/variable/setvar_test.go:126]: No, no, let's make it updatable this time. I'm afraid I'll forget it later.

File: sessionctx/variable/setvar_test.go (hunk 4 above)
Comment (elsa0520) [reply sessionctx/variable/setvar_test.go:134]: OK, just a suggestion. (After all, if all the changeable variables are tested this time, the amount of work will be a bit large.)

File: parser/hintparser.y [hunk 5]
```diff
@@ -562,6 +562,22 @@ Value:
 	{
 		$$ = strconv.FormatUint($1, 10)
 	}
+|	'+' hintIntLit
+	{
+		$$ = strconv.FormatUint($2, 10)
+	}
+|	'-' hintIntLit
```
Comment (elsa0520) [inline parser/hintparser.y:569]: How is this syntax used?

Comment (winoros) [reply parser/hintparser.y:569]: For negative numeric values

Comment (winoros) [reply parser/hintparser.y:569]: [`d140007` (#45893)](https://github.com/pingcap/tidb/pull/45893/commits/d140007d9264c1ba62804d64a542aa1162ecb399)

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by yudongusa — APPROVED** [review]
LGTM for the config approval

**Pull request merged**